### ISSUE TITLE
Derby doesn't throw on invalid template syntax

### ISSUE THREAD

**Issue by aconrad** [derbyjs/derby-parsing#16]
Derby doesn't throw on what seems to be invalid syntax:

```
{{each items as #item)}}
  ...
{{/each}}
```

Notice how there's an accidental closing parenthesis on the first line: `#item)`. This is very confusing because before seeing the error, `#item` is `undefined`, nothing crashes, and it wastes a lot of time to _see_ what/where the problem is.

It would be helpful if Derby threw an error while compiling such template.

**Comment by ericyhwang** [comment]
Thanks for the report! The parser ends up including the closing paren in the loop variable name, which is why it doesn't error out:

```
> template = derbyParsing.createTemplate('{{each items as #item)}}{{/each}}')
> template.content[0].expression
Expression {
  segments: [ 'items' ],
  meta: 
   ExpressionMeta {
     source: 'each items as #item)',
     blockType: 'each',
     isEnd: undefined,
     as: '#item)',
     ... } }
```

Parsing of blocks, including `{{each}}`, is done with a regular expression and not esprima-derby, as the block declarations aren't JS expressions:

https://github.com/derbyjs/derby-parsing/blob/fb62208d7946ffd4df1c6eaa1a06fdc925dd41fe/lib/index.js#L692

After talking to @nateps, we've decided to validate that the parsed variable names are valid identifiers according to esprima-derby.

**Comment by ericyhwang** [comment]
This is interesting... as I was writing tests, I noticed that derby-parsing doesn't enforce that aliases start with `#`, either.

The thing is, Derby doesn't let you _use_ an alias unless it starts with `#`, as that's how it knows to look up an alias. If you use an identifier without a starting `#`, then it's looked up on the model.

As an example, assuming `items = ['a', 'b', 'c']`:

```
{{each items as item)}}
  <li>{{item}}</li>
{{/each}}
```

You'll get three empty `<li>` elements - the each block will iterate over the items just fine, but when you try to use an item, Derby won't locate the item since it's trying to look on the model due to lack of `#`.

I'm thinking we should also add parse-time validation that aliases start with `#`, as aliases lacking `#` aren't currently usable anyways.

**Comment by ericyhwang** [comment]
This is resolved by https://github.com/derbyjs/derby-parsing/pull/17, which introduces parse-time validation on alias names.

It's published as derby-parsing@0.7.4
